doc/flow_counters/routes_flow_counters.md

```diff
@@ -0,0 +1,475 @@
+- Adding route entry shall be automatically bound to counter if counter is enabled and pattern matches
+- Removing route entry shall be automatically unbound if the entry is previously bound
+- Removing route entry shall not trigger route pattern search
+- Decreasing max allowed match count shall not trigger removing of existing bound counters
+- Increasing max allowed match count shall not trigger route pattern search
+
+### Architecture Design
+
+Flow counter shall utilize the following existing infrastructure:
+
+- Generic Counters API - defined in https://github.com/opencomputeproject/SAI/blob/master/doc/SAI-Proposal-Generic-Counters.md and already supported by the SAI layer. This feature shall support binding/unbinding of these Generic counters to/from relevant SONIC objects
+- Flex Counters framework - used for background polling and pushing the statistic information to COUNTER DB for later use (e.g. by CLI). A introduction to flex counter can be found at: https://github.com/Azure/SONiC/pull/858.
+
+![architecture](https://github.com/Junchao-Mellanox/SONiC/blob/route-flow-counter/doc/flow_counters/route_flow_counter_architecture.png).
+
+### High-Level Design
+
+Changes shall be made to sonic-utilities, sonic-swss sub-modules to support this feature. sonic-sairedis needs not change because trap flow counter feature has already implemented all logic for flow counters.
+
+> Note: Code present in this design document is only for demonstrating the design idea, it is not production code.
+
+#### sonic-swss
+
+##### Flex Counter Orch
+
+A new flex counter group shall be added to this class: ROUTE_FLOW_COUNTER. Flex counter orch shall call Route orch to generate/clear flex counter if user enable/disable ROUTE_FLOW_COUNTER, this flow will be described in chapter [Route Orch](#route-orch)
+
+##### FlexCounterManager
+
+FlexCounterManager updates/clears data to FLEX_COUNTER_TABLE so that syncd knows how to query the stats.
+
+A new Counter Type shall be added to FlexCounterManager:
+
+```cpp
+enum class CounterType
+{
+    ...
+    ROUTE_MATCH,
+}
+```
+
+A new entry shall be added to counter_id_field_lookup:
+
+```cpp
+const unordered_map<CounterType, string> FlexCounterManager::counter_id_field_lookup =
+{
+    ...
+    { CounterType::ROUTE_MATCH,  FLOW_COUNTER_ID_LIST },
+}
+```
+
+##### Route Pattern Orch:
+
+A new class Route pattern orch shall be added to orchagent to handle route patterns change in CONFIG_DB. Route pattern orch shall call route orch if router pattern/max allowed match count is configured by user.
+
+##### Route Orch
+
+A ew cache shall be added to Route Orch as data members:
+
+- Cache for those routes which match the prefix pattern and are bound to counters (Bound Cache).
+
+Route Orch shall be extended to handle following cases:
+
+1. Route flow counter enabled
+
+![route-flow-counter-enabled](https://github.com/Junchao-Mellanox/SONiC/blob/route-flow-counter/doc/flow_counters/route_flow_counter_enabled.svg).
```

Comment:
can you use relative path so that you can remove your user specific info in the url?